src/machine.rs

```diff
@@ -53,17 +52,17 @@ where
         debug_assert!(node.all(|id| id == Id::from(0)));
         debug_assert!(eclass.nodes.windows(2).all(|w| w[0] < w[1]));
         let mut start = eclass.nodes.binary_search(node).unwrap_or_else(|i| i);
-        let discrim = std::mem::discriminant(node);
+        let discrim = node.discriminant();
         while start > 0 {
-            if std::mem::discriminant(&eclass.nodes[start - 1]) == discrim {
+            if eclass.nodes[start - 1].discriminant() == discrim {
```

Comment:
I wasn't sure about this check here. Why is it expected that nodes with equal discriminant will be adjacent? It doesn't seem like the nodes are sorted by discriminant (the Discriminant doesn't implement Ord).